library/core/src/num/niche_types.rs

```diff
@@ -160,6 +172,15 @@ define_valid_range_type! {
     pub struct I64NotAllOnes(i64 as u64 in 0..=0xffff_ffff_ffff_fffe);
 }
 
+#[cfg(not(bootstrap))]
+define_valid_range_type! {
+    pub struct U32NotAllOnes(u32 as u32 in 0..=0xffff_fffe);
+    pub struct I32NotAllOnes(i32 as u32 in -0x7fff_fffe..);
```

Comment:
`-0x7fff_fffe..` is very much the wrong range since it excludes `-0x8000_0000..=-0x7fff_ffff` rather than `-1`

imo this should be `..=-2 | 0..`

Reply:
Oh lmao thx.

Reply:
I would probably also change the macro input to be:
(remove the parenthesis around the `|` patterns if not needed)
```rust
define_valid_range_type! {
    #[uint = u32]
    pub struct U32NotAllOnes(u32 in 0..=0xffff_fffe);
    #[uint = u32]
    pub struct I32NotAllOnes(i32 in (..=-2 | 1..));

    #[uint = u64]
    pub struct U64NotAllOnes(u64 in 0..=0xffff_ffff_ffff_fffe);
    #[uint = u64]
    pub struct I64NotAllOnes(i64 in (..=-2 | 1..));
}
```

Reply:
How do you like the latest version? Unfortunately I had to duplicate the validity check, because there is no way to ask a pattern type for its valid range/check the range in a generic way at present (can't write generic pattern types without an incomplete feature gate which causes all kinds of problems, so it can't be non-incomplete 😆 )

Reply:
from just reading the library changes, it looks good enough to me (except the 1 nit I commented on). though it did occur to me, wouldn't just using this in `new()` work? it *is* a pattern, built-in pattern matching should work...
```rust
impl WrapperTypeOverU32 {
    pub fn new(input: u32) -> Option<Self> {
        if let $pat = input {
            Some(...)
        } else {
            None
        }
    }
}
```

Reply:
🤦 thank you.